src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,719 @@
+        std::visit(match {
+            [&](const CKeyID& keyId) {
+                tfvk = pwalletMain->GetLegacyAccountKey().ToAccountPubKey();
+            },
+            [&](const CScriptID& keyId) {
+                tfvk = pwalletMain->GetLegacyAccountKey().ToAccountPubKey();
+            },
+            [&](const UnifiedAddress& ua) {
+                // This is safe because spending key checks will have ensured that we
+                // have a UFVK corresponding to this address, and transparent UTXOs will
+                // not have been selected if the UFVK does not contain a transparent key.
+                auto ufvk = pwalletMain->GetUFVKForAddress(ua);
+                tfvk = ufvk.value().GetTransparentKey().value();
+            },
+            [&](const UnifiedFullViewingKey& ufvk) {
+                // Transparent UTXOs will not have been selected if the UFVK does not contain
+                // a transparent key.
+                tfvk = ufvk.GetTransparentKey().value();
+            },
+            [&](const AccountZTXOPattern& acct) {
+                if (acct.GetAccountId() == ZCASH_LEGACY_ACCOUNT) {
+                    tfvk = pwalletMain->GetLegacyAccountKey().ToAccountPubKey();
+                } else {
+                    // By definition, we have a UFVK for every known account.
+                    auto ufvk = pwalletMain->GetUnifiedFullViewingKeyByAccount(acct.GetAccountId()).value();
+                    // Transparent UTXOs will not have been selected if the UFVK does not contain
+                    // a transparent key.
+                    tfvk = ufvk.GetTransparentKey().value();
+                }
+            },
+            [&](const auto& other) {
+                throw std::runtime_error("SelectOVKs: Selector cannot select transparent UTXOs.");
+            }
+        }, selector.GetPattern());
+        assert(tfvk.has_value());
+
+        auto ovks = tfvk.value().GetOVKsForShielding();
+        internalOVK = ovks.first;
+        externalOVK = ovks.second;
+    } else if (!spendable.sproutNoteEntries.empty()) {
+        // use the legacy transparent account OVKs when sending from Sprout
+        auto tfvk = pwalletMain->GetLegacyAccountKey().ToAccountPubKey();
+        auto ovks = tfvk.GetOVKsForShielding();
+        internalOVK = ovks.first;
+        externalOVK = ovks.second;
+    } else {
+        // This should be unreachable; it is left in place as a guard to ensure
+        // that when new input types are added to SpendableInputs in the future
+        // that we do not accidentally return the all-zeros OVK.
+        throw std::runtime_error("No spendable inputs.");
+    }
+
+    return std::make_pair(internalOVK, externalOVK);
+}
+
+PrivacyPolicy TransactionEffects::GetRequiredPrivacyPolicy() const
+{
+    if (!spendable.utxos.empty()) {
+        std::set<std::optional<AddressUFVKMetadata>> uas;
+        for (const auto& utxo : spendable.utxos) {
+            uas.insert(utxo.second);
+        }
```

Comment:
Does it now make sense, with the strengthening of `AllowTransparentSenders` and `AllowLinkingAccountAddresses`, to now just collect the set of taddrs?

Reply:
I would like to avoid changing the semantics in this PR. Since this particular change does cascade through a few things, is contained in its own commit, and is part of the new-in-this-PR `GetRequiredPrivacyPolicy`, I would personally be fine with omitting it here (leaving the gap in coverage for `ALAA`/ `NoPrivacy` that existed in earlier incarnations of this PR) then having a second (or 14th or something) PR that introduces the tighter semantics for `AllowRevealedSenders`.

Reply:
See https://github.com/zcash/zcash/pull/6377#discussion_r1130017108